src/twin/system_info.rs

```diff
@@ -118,4 +204,112 @@ impl SystemInfo {
         .await
         .context("report: send")
     }
+
+    async fn metrics_handler(&mut self) -> Result<()> {
+        let time = match time::OffsetDateTime::now_utc().format(&Rfc3339) {
+            Ok(time) => time,
+            Err(e) => {
+                warn!("timestamp could not be generated: {e}");
+                return Ok(());
+            }
+        };
+
+        self.sysinfo_components.refresh_list();
+        self.sysinfo_memory.refresh_cpu_usage();
+        self.sysinfo_memory.refresh_memory();
+        self.sysinfo_disk.refresh();
+
+        let Some(hostname) = System::host_name() else {
+            warn!("hostname could not be read");
+            return Ok(());
+        };
+
+        let mut disk_total = 0;
+        let mut disk_used = 0;
+        for disk in self.sysinfo_disk.list() {
+            if disk.name().to_str() == Some("/dev/omnect/data") {
+                disk_total = disk.total_space();
+                disk_used = disk.total_space() - disk.available_space();
+                break;
+            }
+        }
+
+        let mut metric_list = vec![
+            Metric::new(
+                time.clone(),
+                "cpu_usage",
+                self.sysinfo_memory.global_cpu_usage() as f64,
+                hostname.clone(),
+                None,
+            ),
+            Metric::new(
+                time.clone(),
+                "memory_used",
+                self.sysinfo_memory.used_memory() as f64,
+                hostname.clone(),
+                None,
+            ),
+            Metric::new(
+                time.clone(),
+                "memory_total",
+                self.sysinfo_memory.total_memory() as f64,
+                hostname.clone(),
+                None,
+            ),
+            Metric::new(
+                time.clone(),
+                "disk_used",
+                disk_used as f64,
+                hostname.clone(),
+                None,
+            ),
+            Metric::new(
+                time.clone(),
+                "disk_total",
+                disk_total as f64,
+                hostname.clone(),
+                None,
+            ),
+        ];
+
+        for component in self.sysinfo_components.iter() {
+            metric_list.push(Metric::new(
+                time.clone(),
+                "temp",
+                component.temperature() as f64,
+                hostname.clone(),
+                Some(format!("{}", component.label())),
+            ));
+        }
+
+        match serde_json::to_vec(&metric_list) {
+            Ok(json) => {
+                match IotMessage::builder()
+                    .set_body(json)
+                    .set_content_type("application/json")
+                    .set_content_encoding("utf-8")
+                    .set_output_queue("metrics")
+                    .build()
+                {
+                    Ok(msg) => {
+                        let Some(tx) = &self.tx_outgoing_message else {
```

Comment:
this check should be moved to top of function

Reply:
done